### ISSUE TITLE
a few questions and issues about using libpnet

### ISSUE THREAD

**Issue by little-dude** [libpnet/libpnet#134]
## Checksum

When trying to implement icmp packets, I've added one `#[packet]` per icmp
type. Checksum calculation is the same for all those packets, but the function
I wrote for it only accepts one type of packets:

```rust
/// Calculates the checksum of an ICMP packet
pub fn checksum(packet: &IcmpPacket) -> u16be {
    use packet::Packet;

    let mut sum = 0u32;
    let mut i = 0;
    while i < packet.packet().len() {
        let word = (packet.packet()[i] as u32) << 8 | packet.packet()[i + 1] as u32;
        sum = sum + word;
        i = i + 2;
    }
    while sum >> 16 != 0 {
        sum = (sum >> 16) + (sum & 0xFFFF);
    }

    !sum as u16
}
```

I didn't try too much but I guess I could solve this by implementing a trait
`IcmpChecksum` and use `#[derive(IcmpChecksum)]` on all the icmp packets.

## Struggling with ownership concepts

Coming from python, I still struggling with some basic things. Here is what I would have been able to do but could get working.

### nested `new`

When I `know` a buffer represents an icmp packet, I'd like to be able to do
this:

```rust
let capture = &testutils::read_capture("echo_request");
// let say `capture[0]` is a Vec<u8> representing an icmp echo request packet
let icmp_packet =
    EchoRequest::new(
        Ipv4Packet::new(
            EthernetPacket::new($capture[0][..])
            .unwrap()
            .payload())
        .unwrap()
        .payload())
    .unwrap();
```

But this does not work because the ethernet packet does not live long enough.
So I have to go with this version:

```rust
let capture = &testutils::read_capture("echo_request");
// let say `capture[0]` is a Vec<u8> representing an icmp echo request packet
let ethernet = EthernetPacket::new($capture[0][..]).unwrap();
let ip = Ipv4Packet::new(ethernet.payload()).unwrap();
let echo_request =  EchoRequest::new(ip.payload()).unwrap();
```

### returning a packet

when writing tests for icmp packets I wanted to be able to write something like
this (I know, it saves only two lines, but these are two lines I would have to
write all the time):

```rust
fn get_icmp_buf(packet_buf: &[u8]) -> &[u8] {
    let eth = EthernetPacket::new(packet_buf).unwrap();
    let ip = Ipv4Packet::new(eth.payload()).unwrap():
    ip.payload();
}

// then later:

let capture = testutils::read_capture("time_exceeded");
let time_exceeded_packet = TimeExceeded::new(get_icmp_buf(&capture[0][..])).unwrap();
```

But the compiler complains so I ended up copying the data:

```rust
fn get_ip_payload(packet_bytes: &[u8]) -> Vec<u8> {
    let eth_packet = EthernetPacket::new(packet_bytes).unwrap();
    let ip_packet = Ipv4Packet::new(eth_packet.payload()).unwrap();
    let mut bytes = vec![];
    for byte in ip_packet.payload() {
        bytes.push(*byte);
    }
    bytes
}
```

For the same reason, I cannot make a function that returns an ip packet

```rust
fn get_ip_pakcet(packet_bytes: &[u8]) -> Ipv4Packet {
    let eth_packet = EthernetPacket::new(packet_bytes).unwrap();
    Ipv4Packet::new(eth_packet.payload()).unwrap();
}
```

## panic! 

when an ip packet does not have any option in the header the following seems to `panic!`:

```rust
ip_packet.get_options_raw()
 thread 'packet::icmp::tests::destination_unreachable' panicked at 'slice index starts at 20 but ends at 0', src/libcore/slice.rs:578
```

## bump version

I don't really know how versioning works with rust, but I had to change pnet version to `0.5.2` to make it compile. Should I open a PR for this?
https://github.com/little-dude/libpnet/commit/c9bbacd6c02126d863cd558f2dfc41d07964fcf8#diff-80398c5faae3c069e4e6aa2ed11b28c0L3

**Comment by mrmonday** [comment]
I'm going to bounce between this and a few other things, so I'll reply as a series of comments over the next few hours rather than all at once:

##### Checksum

You should take a look at the work being done by @david415 in pulls #98 and #105. They were looking to break out the checksum functionality to remove the duplication. I don't know if they're still going with it or not. I think it's important it gets modularised at some point, particularly given the increasing duplication. For now it's fine that you've copy/pasted it. I can't remember what validation (if any) is done for checksum functions, but you could try something like:

```rust
pub fn checksum<T : Packet>(packet: &T) -> u16be {
    ...
}
```

Or alternatively write a small macro which generates the functions so you don't have to copy/paste it.

##### Version Bump

It's weird that you have to change the version, that shouldn't be necessary. You can read about Rust/Cargo's versioning system [here](http://semver.org/). Since libpnet is still pre-1.0, the numbers are fairly meaningless (see point 4 of the semantic versioning spec). I've been trying to stick to `0.X.Y` where X is updated for new features and breaking changes, and Y is updated for bug fixes/updates to work with the latest Rust nightly. Don't worry about it for now - we'll talk about it as a final step for when your changes are ready for merging :)

**Comment by mrmonday** [comment]
##### Nested `new`

I find the second thing you wrote to be a lot more readable, so I'm somewhat glad the compiler doesn't accept it! Here's what's happening (roughly) - the definition of `new` for the `*Packet` types is:

```rust
fn new<'p>(packet: &'p [u8]) -> Option<EthernetPacket<'p>>
```

Obviously replacing `EthernetPacket` with whichever is appropriate.

What this says is "for some lifetime `'p`, given a slice of `u8`s, which lives for as long as `'p`, this function will return either `None`, or some `EthernetPacket`, which is valid for the lifetime `'p`".

The lifetime of `'p` is inferred when you call `new()`. When you write it as nested calls, Rust infers that `'p` needs to live for as long as the outer call, which means it is then not valid for the entire function. Using a let binding lets Rust infer the correct lifetime.

##### Returning a Packet

What error messages are you getting for that code? I've written code which does similar things before, it should be possible to do what you're trying to achieve. From a quick glance it looks like it might be lifetime issues, similar to those you were experiencing when trying to nest calls to `new()`? If so, breaking things into separate `let` bindings should help.

**Comment by mrmonday** [comment]
##### panic!

It looks like you've uncovered two bugs here, could you file issues for them? The test cases are as follows:

The issue you encountered:

```rust
extern crate pnet;

use pnet::packet::ipv4::MutableIpv4Packet;

fn main() {
    let buffer: &mut [u8] = &mut [0; 64];
    let mut packet = MutableIpv4Packet::new(buffer).unwrap();
    packet.set_header_length(5);

    println!("{:?}", packet.get_options_raw());
}
```

Another issue I encountered whilst trying to reproduce:

```rust
extern crate pnet;

use pnet::packet::ipv4::Ipv4Packet;

fn main() {
    let buffer: &[u8] = &[0; 64];
    let packet = Ipv4Packet::new(buffer).unwrap();

    println!("{:?}", packet.get_options_raw());
}
```

Ultimately I want libpnet/pnet_macros to get to the point where it is possible to guarantee at compile time that none of it panics - it's definitely not in that state currently though.

##### And finally...

I think I've responded to everything you've asked now - did I miss anything? It was harder than I expected to keep track of what I'd answered, did a lot of jumping around, and obviously I didn't do it all at the same time.

**Comment by david415** [comment]
yes we should really make the proper abstractions here so we don't have code duplication for packet checksums... i think that this was not a good first rust project for me as i lost steam trying to debug the bit fiddling. i'd really just like for the library to exist in a complete enough form so that i can write scanning tools and protocol analysis tools like the infamous honeybadger. https://github.com/david415/HoneyBadger ;-p

**Comment by mrmonday** [comment]
Great to have you back, @david415!

I agree, we need the proper abstractions in there - this was my first *real* networking project, which is why there's no abstraction in place in the first place - I didn't realise how ubiquitous the internet checksum was across different packet types. It didn't even occur to me that I'd written the same code multiple times until you pointed it out!

Bit fiddling is a slow, arduous, and difficult task - that's how libpnet_macros came about - as an attempt to remove the need to do it in the majority of networking code. With that in mind, it becomes somewhat obvious that we need a proper abstraction for checksums, since you don't want to have to write that code every time.

I'd love libpnet to be the one-stop-shop for low level networking - obviously there's some way to go though. Maybe it's time to start building a roadmap of what needs to happen for it to be a generally useful tool.

As the author of HoneyBadger, @david415, what's missing from libpnet to make it useful/easy to use as a TCP stream analysis tool?

**Comment by mrmonday** [comment]
I've opened a milestone for things that need doing before libpnet 1.0: https://github.com/libpnet/libpnet/milestones/libpnet%201.0

Pretty much anything you come up with stands a good chance of getting on that list :)

**Comment by mrmonday** [comment]
I believe pull #153 should fix the panics occurring here. Just leaving a reminder to check.

**Comment by david415** [comment]
hi! Sorry I've been away for so long. Does this mean the checksum branch is fixed? Or does it need to be rewritten to work with the latest code?

I need a TCP encoding/decoding API and it will be easy to write TCP stream analysis tools. I wanted to make checksums modular before solving for TCP.

**Comment by mrmonday** [comment]
I've been really busy for the past couple of months too!

There are 4 pull requests for checksums open between yourself and @little-dude, I don't know what the statuses of them are, or which one has the latest code. It looks like it might have been #150?

I'd love to get the generic checksum support reviewed and merged in - it's the longest standing pull request (ever!) for libpnet, and it's really important functionality. If you don't mind following it up, I'd be happy to review the latest branch and see if we can finally merge it.
